File: fabric/src/main/java/com/eliasnvx/krylix/fabric/client/LeaderboardScreen.java
package com.eliasnvx.krylix.fabric.client;

import net.minecraft.client.renderer.RenderPipelines;
import com.eliasnvx.krylix.fabric.network.FabricNetworkPackets.PlayerStatEntry;
import com.mojang.authlib.GameProfile;
import com.mojang.blaze3d.systems.RenderSystem;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.client.multiplayer.PlayerInfo;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.Identifier;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class LeaderboardScreen extends Screen {
    private enum Mode { PLAYERS, MOB_KILLS }

    private final int panelWidth = 340;
    private final int panelHeight = 260;
    private final int rowHeight = 22;
    private final int headerHeight = 58;
    private final int footerHeight = 18;
    private final int avatarSize = 16;
    private final int tabHeight = 16;

    private final int rankColumnEnd = 40;
    private final int avatarColumnX = 46;

    private final int pageSize = 60;

    private Mode mode = Mode.PLAYERS;
    private int page = 0;
    private int scrollOffset = 0;
    private int[] prevButtonRect = null;
    private int[] nextButtonRect = null;

    private final List<PlayerStatEntry> playerEntries = PlayerStatsClient.sortedByKills();
    private final List<PlayerStatEntry> mobKillEntries = PlayerStatsClient.sortedByMobKills();
    private final Map<String, Identifier> headTextures = new HashMap<>();

    public LeaderboardScreen() {
        super(Component.translatable("screen.krylix.leaderboard"));
    }

    @Override
    protected void init() {
        for (PlayerStatEntry entry : playerEntries) {
            headTextures.computeIfAbsent(entry.uuid(), k -> resolveSkin(entry));
        }
        for (PlayerStatEntry entry : mobKillEntries) {
            headTextures.computeIfAbsent(entry.uuid(), k -> resolveSkin(entry));
        }
    }

    private Identifier resolveSkin(PlayerStatEntry entry) {
        try {
            UUID uuid = UUID.fromString(entry.uuid());
            if (minecraft != null && minecraft.getConnection() != null) {
                PlayerInfo info = minecraft.getConnection().getPlayerInfo(uuid);
                if (info != null && info.getSkin() != null) {
                    return info.getSkin().body().texturePath();
                }
                return net.minecraft.client.resources.DefaultPlayerSkin.get(uuid).body().texturePath();
            }
        } catch (Exception ignored) { }
        return null;
    }

    private int panelX() { return (width - panelWidth) / 2; }
    private int panelY() { return (height - panelHeight) / 2; }
    private int listTop() { return panelY() + headerHeight; }
    private int listBottom() { return panelY() + panelHeight - footerHeight; }
    private int listHeight() { return listBottom() - listTop(); }

    private List<PlayerStatEntry> fullEntries() {
        return mode == Mode.PLAYERS ? playerEntries : mobKillEntries;
    }

    private int totalPages() {
        return Math.max(1, (fullEntries().size() + pageSize - 1) / pageSize);
    }

    private List<PlayerStatEntry> pageEntries() {
        List<PlayerStatEntry> full = fullEntries();
        int from = Math.max(0, Math.min(full.size(), page * pageSize));
        int to = Math.max(0, Math.min(full.size(), (page + 1) * pageSize));
        return full.subList(from, to);
    }

    private int rowCount() {
        return pageEntries().size() + (totalPages() > 1 ? 1 : 0);
    }

    private int maxScroll() {
        return Math.max(0, rowCount() * rowHeight - listHeight());
    }

    @Override
    public void extractRenderState(GuiGraphicsExtractor guiGraphics, int mouseX, int mouseY, float partialTick) {
        super.extractRenderState(guiGraphics, mouseX, mouseY, partialTick);

        int px = panelX();
        int py = panelY();

        HudRender.rounded(guiGraphics, px, py, panelWidth, panelHeight, 6, () -> {
            guiGraphics.fill(px, py, px + panelWidth, py + panelHeight, 0xE6101014);
        });

        guiGraphics.centeredText(font, title, px + panelWidth / 2, py + 8, 0xFFFFFF);

        int closeX = px + panelWidth - 18;
        int closeY = py + 7;
        boolean closeHovered = mouseX >= closeX && mouseX <= closeX + 12 && mouseY >= closeY && mouseY <= closeY + 12;
        guiGraphics.text(font, "x", closeX + 3, closeY + 2, closeHovered ? 0xFFFFFF : 0x999999);

        renderTabs(guiGraphics, mouseX, mouseY, px, py);

        int headerY = py + headerHeight - 20;
        String statsHeaderText = mode == Mode.PLAYERS ?
                Component.translatable("krylix.leaderboard.header_kd").getString() :
                Component.translatable("krylix.leaderboard.tab_mob_kills").getString();
        String rankHeaderText = Component.translatable("krylix.leaderboard.header_rank").getString();
        
        guiGraphics.text(font, rankHeaderText, px + 12, headerY, 0x888888);
        guiGraphics.text(font, Component.translatable("krylix.leaderboard.header_player").getString(), px + avatarColumnX, headerY, 0x888888);
        int statsHeaderWidth = font.width(statsHeaderText);
        guiGraphics.text(font, statsHeaderText, px + panelWidth - statsHeaderWidth - 12, headerY, 0x888888);
        guiGraphics.fill(px + 8, py + headerHeight - 4, px + panelWidth - 8, py + headerHeight - 3, 0x40FFFFFF);

        if (fullEntries().isEmpty()) {
            String emptyKey = mode == Mode.PLAYERS ? "screen.krylix.leaderboard.empty_players" : "screen.krylix.leaderboard.empty_mob_kills";
            String emptyText = Component.translatable(emptyKey).getString();
            guiGraphics.centeredText(font, emptyText, px + panelWidth / 2, listTop() + listHeight() / 2 - 4, 0xAAAAAA);
        } else {
            renderRows(guiGraphics, mouseX, mouseY);
        }

        String footerText = Component.translatable("krylix.leaderboard.footer", fullEntries().size()).getString();
        guiGraphics.centeredText(font, footerText, px + panelWidth / 2, py + panelHeight - 14, 0x777777);
    }

    private int[] tabRect(int index, int px, int py) {
        int tabWidth = 90;
        int gap = 6;
        int totalWidth = tabWidth * 2 + gap;
        int startX = px + (panelWidth - totalWidth) / 2;
        int x = startX + index * (tabWidth + gap);
        int y = py + 20;
        return new int[]{x, y, tabWidth, tabHeight};
    }

    private void renderTabs(GuiGraphicsExtractor guiGraphics, int mouseX, int mouseY, int px, int py) {
        renderTab(guiGraphics, mouseX, mouseY, tabRect(0, px, py), Component.translatable("krylix.leaderboard.tab_pvp").getString(), Mode.PLAYERS);
        renderTab(guiGraphics, mouseX, mouseY, tabRect(1, px, py), Component.translatable("krylix.leaderboard.tab_mob_kills").getString(), Mode.MOB_KILLS);
    }

    private void renderTab(GuiGraphicsExtractor guiGraphics, int mouseX, int mouseY, int[] rect, String label, Mode tabMode) {
        int x = rect[0], y = rect[1], w = rect[2], h = rect[3];
        boolean active = mode == tabMode;
        boolean hovered = mouseX >= x && mouseX <= x + w && mouseY >= y && mouseY <= y + h;
        
        HudRender.rounded(guiGraphics, x, y, w, h, 3, () -> {
            int bg = active ? 0xCC3A7BD5 : (hovered ? 0x30FFFFFF : 0x20FFFFFF);
            guiGraphics.fill(x, y, x + w, y + h, bg);
        });
        
        int textColor = active ? 0xFFFFFF : 0xAAAAAA;
        int textWidth = font.width(label);
        guiGraphics.text(font, label, x + (w - textWidth) / 2, y + (h - 8) / 2, textColor);
    }

    private void renderRows(GuiGraphicsExtractor guiGraphics, int mouseX, int mouseY) {
        int px = panelX();
        int top = listTop();
        int bottom = listBottom();
        prevButtonRect = null;
        nextButtonRect = null;

        HudRender.rounded(guiGraphics, px + 6, top, panelWidth - 12, bottom - top, 4, () -> {
            String localUuid = minecraft != null && minecraft.player != null ? minecraft.player.getUUID().toString() : null;
            List<PlayerStatEntry> entries = pageEntries();

            int y = top - scrollOffset;
            for (int i = 0; i < entries.size(); i++) {
                if (y + rowHeight >= top && y <= bottom) {
                    PlayerStatEntry entry = entries.get(i);
                    boolean rowHovered = mouseX >= px + 6 && mouseX <= px + panelWidth - 6 && mouseY >= y && mouseY <= y + rowHeight;
                    boolean isSelf = entry.uuid().equals(localUuid);
                    
                    int bgColor = 0;
                    if (rowHovered) bgColor = 0x33FFFFFF;
                    else if (isSelf) bgColor = 0x2255AAFF;
                    else if (i % 2 == 0) bgColor = 0x18FFFFFF;
                    
                    if (bgColor != 0) {
                        guiGraphics.fill(px + 6, y, px + panelWidth - 6, y + rowHeight, bgColor);
                    }

                    renderRow(guiGraphics, entry, page * pageSize + i, px, y);
                }
                y += rowHeight;
            }

            if (totalPages() > 1 && y + rowHeight >= top && y <= bottom) {
                renderPaginationRow(guiGraphics, mouseX, mouseY, px, y);
            }
        });
    }

    private void renderPaginationRow(GuiGraphicsExtractor guiGraphics, int mouseX, int mouseY, int px, int y) {
        int btnSize = 16;
        int gap = 8;
        String pageText = Component.translatable("krylix.leaderboard.page", page + 1, totalPages()).getString();
        int pageTextWidth = font.width(pageText);
        int totalWidth = btnSize + gap + pageTextWidth + gap + btnSize;
        int startX = px + (panelWidth - totalWidth) / 2;
        int btnY = y + (rowHeight - btnSize) / 2;

        boolean hasPrev = page > 0;
        boolean hasNext = page < totalPages() - 1;

        prevButtonRect = new int[]{startX, btnY, btnSize, btnSize};
        nextButtonRect = new int[]{startX + btnSize + gap + pageTextWidth + gap, btnY, btnSize, btnSize};

        renderPageButton(guiGraphics, mouseX, mouseY, prevButtonRect, false, hasPrev);
        renderPageButton(guiGraphics, mouseX, mouseY, nextButtonRect, true, hasNext);
        guiGraphics.text(font, pageText, startX + btnSize + gap, y + (rowHeight - 8) / 2, 0xCCCCCC);
    }

    private void renderPageButton(GuiGraphicsExtractor guiGraphics, int mouseX, int mouseY, int[] rect, boolean pointRight, boolean enabled) {
        int x = rect[0], y = rect[1], w = rect[2], h = rect[3];
        boolean hovered = enabled && mouseX >= x && mouseX <= x + w && mouseY >= y && mouseY <= y + h;
        
        HudRender.rounded(guiGraphics, x, y, w, h, 3, () -> {
            int bg = !enabled ? 0x15FFFFFF : (hovered ? 0x40FFFFFF : 0x25FFFFFF);
            guiGraphics.fill(x, y, x + w, y + h, bg);
        });
        
        int color = enabled ? 0xFFFFFFFF : 0xFF555555;
        renderArrow(guiGraphics, x, y, w, h, pointRight, color);
    }

    private void renderArrow(GuiGraphicsExtractor guiGraphics, int x, int y, int w, int h, boolean pointRight, int color) {
        int arrowW = 4;
        int arrowH = 7;
        int originX = x + (w - arrowW) / 2;
        int originY = y + (h - arrowH) / 2;
        int mid = arrowH / 2;
        for (int row = 0; row < arrowH; row++) {
            int d = Math.abs(row - mid);
            int length = Math.max(1, arrowW - d);
            int rowY = originY + row;
            if (pointRight) {
                guiGraphics.fill(originX, rowY, originX + length, rowY + 1, color);
            } else {
                guiGraphics.fill(originX + (arrowW - length), rowY, originX + arrowW, rowY + 1, color);
            }
        }
    }

    private void renderRow(GuiGraphicsExtractor guiGraphics, PlayerStatEntry entry, int rank, int px, int y) {
        int textY = y + (rowHeight - 8) / 2;
        int avatarY = y + (rowHeight - avatarSize) / 2;

        String rankText = "#" + (rank + 1);
        int rankX = px + rankColumnEnd - font.width(rankText);
        guiGraphics.text(font, rankText, rankX, textY, 0xAAAAAA);

        int avatarX = px + avatarColumnX;
        Identifier texture = headTextures.get(entry.uuid());
        
        HudRender.rounded(guiGraphics, avatarX, avatarY, avatarSize, 2, () -> {
            if (texture != null) {
                guiGraphics.blit(RenderPipelines.GUI_TEXTURED, texture, avatarX, avatarY, 8.0f, 8.0f, 8, 8, 64, 64, avatarSize, avatarSize);
                guiGraphics.blit(RenderPipelines.GUI_TEXTURED, texture, avatarX, avatarY, 40.0f, 8.0f, 8, 8, 64, 64, avatarSize, avatarSize);
            } else {
                guiGraphics.fill(avatarX, avatarY, avatarX + avatarSize, avatarY + avatarSize, 0xFF555555);
            }
        });

        guiGraphics.text(font, entry.name(), avatarX + avatarSize + 6, textY, 0xFFFFFF);

        String statsText;
        if (mode == Mode.PLAYERS) {
            double kd = entry.deaths() == 0 ? entry.kills() : (double) entry.kills() / entry.deaths();
            statsText = entry.kills() + "K / " + entry.deaths() + "D  (" + String.format("%.2f", kd) + ")";
        } else {
            statsText = Component.translatable("krylix.leaderboard.mob_kills_count", entry.mobKills()).getString();
        }
        int statsWidth = font.width(statsText);
        guiGraphics.text(font, statsText, px + panelWidth - statsWidth - 12, textY, 0xCCCCCC);
    }

    @Override
    public boolean mouseScrolled(double mouseX, double mouseY, double scrollX, double scrollY) {
        scrollOffset = Math.max(0, Math.min(maxScroll(), scrollOffset - (int) (scrollY * rowHeight * 1.5)));
        return true;
    }

    @Override
    public boolean mouseClicked(net.minecraft.client.input.MouseButtonEvent event, boolean doubleClick) {
        double mouseX = event.x();
        double mouseY = event.y();
        int px = panelX();
        int py = panelY();
        int closeX = px + panelWidth - 18;
        int closeY = py + 7;
        
        if (mouseX >= closeX && mouseX <= closeX + 12 && mouseY >= closeY && mouseY <= closeY + 12) {
            this.onClose();
            return true;
        }

        int[] t0 = tabRect(0, px, py);
        if (mouseX >= t0[0] && mouseX <= t0[0] + t0[2] && mouseY >= t0[1] && mouseY <= t0[1] + t0[3]) {
            mode = Mode.PLAYERS;
            page = 0;
            scrollOffset = 0;
            return true;
        }
        
        int[] t1 = tabRect(1, px, py);
        if (mouseX >= t1[0] && mouseX <= t1[0] + t1[2] && mouseY >= t1[1] && mouseY <= t1[1] + t1[3]) {
            mode = Mode.MOB_KILLS;
            page = 0;
            scrollOffset = 0;
            return true;
        }

        if (prevButtonRect != null) {
            int[] r = prevButtonRect;
            if (page > 0 && mouseX >= r[0] && mouseX <= r[0] + r[2] && mouseY >= r[1] && mouseY <= r[1] + r[3]) {
                page--;
                scrollOffset = 0;
                return true;
            }
        }
        if (nextButtonRect != null) {
            int[] r = nextButtonRect;
            if (page < totalPages() - 1 && mouseX >= r[0] && mouseX <= r[0] + r[2] && mouseY >= r[1] && mouseY <= r[1] + r[3]) {
                page++;
                scrollOffset = 0;
                return true;
            }
        }

        return super.mouseClicked(event, doubleClick);
    }

    @Override
    public boolean isPauseScreen() {
        return false;
    }
}

File: forge/src/main/kotlin/com/eliasnvx/krylix/forge/client/LeaderboardScreen.kt
package com.eliasnvx.krylix.forge.client

import com.eliasnvx.krylix.forge.network.NetworkPackets.PlayerStatsSyncPacket.PlayerStatEntry
import com.mojang.authlib.GameProfile
import com.mojang.blaze3d.systems.RenderSystem
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.Screen
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation
import java.util.UUID
import kotlin.math.max

/**
 * GUI-экран рейтинга — тёмная полупрозрачная скруглённая панель со скроллящимся списком, в стиле
 * магазинных/каталожных GUI, а не голый список в чате. Две вкладки, обе про игроков: PvP-рейтинг
 * (килы/смерти игроков друг друга) и рейтинг по личным убийствам мобов (кто из игроков убил
 * больше мобов — не список видов мобов).
 */
class LeaderboardScreen : Screen(Component.translatable("screen.krylix.leaderboard")) {

    private enum class Mode { PLAYERS, MOB_KILLS }

    private val panelWidth = 340
    private val panelHeight = 260
    private val rowHeight = 22
    private val headerHeight = 58
    private val footerHeight = 18
    private val avatarSize = 16
    private val tabHeight = 16

    /** Правая граница колонки ранга ("#150" и т.п. правым краем упирается сюда, не наезжая на аватар) */
    private val rankColumnEnd = 40
    private val avatarColumnX = 46

    private val pageSize = 60

    private var mode = Mode.PLAYERS
    private var page = 0
    private var scrollOffset = 0
    private var prevButtonRect: IntArray? = null
    private var nextButtonRect: IntArray? = null

    private val playerEntries: List<PlayerStatEntry> = PlayerStatsClient.sortedByKills()
    private val mobKillEntries: List<PlayerStatEntry> = PlayerStatsClient.sortedByMobKills()
    private val headTextures = mutableMapOf<String, ResourceLocation?>()

    override fun init() {
        (playerEntries + mobKillEntries).forEach { entry -> headTextures.getOrPut(entry.uuid) { resolveSkin(entry) } }
    }

    private fun resolveSkin(entry: PlayerStatEntry): ResourceLocation? {
        val uuid = runCatching { UUID.fromString(entry.uuid) }.getOrNull() ?: return null
        val connection = minecraft?.connection
        connection?.getPlayerInfo(uuid)?.skinLocation?.let { return it }
        return runCatching { minecraft?.skinManager?.getInsecureSkinLocation(GameProfile(uuid, entry.name)) }.getOrNull()
    }

    private fun panelX() = (width - panelWidth) / 2
    private fun panelY() = (height - panelHeight) / 2
    private fun listTop() = panelY() + headerHeight
    private fun listBottom() = panelY() + panelHeight - footerHeight
    private fun listHeight() = listBottom() - listTop()

    private fun fullEntries() = if (mode == Mode.PLAYERS) playerEntries else mobKillEntries
    private fun totalPages() = max(1, (fullEntries().size + pageSize - 1) / pageSize)
    private fun pageEntries(): List<PlayerStatEntry> {
        val full = fullEntries()
        val from = (page * pageSize).coerceIn(0, full.size)
        val to = ((page + 1) * pageSize).coerceIn(0, full.size)
        return full.subList(from, to)
    }

    /** Строк в скролл-зоне: сами записи + виртуальная строка с кнопками страниц (если страниц > 1) */
    private fun rowCount() = pageEntries().size + if (totalPages() > 1) 1 else 0
    private fun maxScroll() = max(0, rowCount() * rowHeight - listHeight())

    override fun render(guiGraphics: GuiGraphics, mouseX: Int, mouseY: Int, partialTick: Float) {
        renderBackground(guiGraphics)

        val px = panelX()
        val py = panelY()

        HudRender.rounded(guiGraphics, px, py, panelWidth, panelHeight, 6) {
            guiGraphics.fill(px, py, px + panelWidth, py + panelHeight, 0xE6101014.toInt())
        }

        guiGraphics.drawCenteredString(font, title, px + panelWidth / 2, py + 8, 0xFFFFFF)

        val closeX = px + panelWidth - 18
        val closeY = py + 7
        val closeHovered = mouseX in closeX..(closeX + 12) && mouseY in closeY..(closeY + 12)
        guiGraphics.drawString(font, "x", closeX + 3, closeY + 2, if (closeHovered) 0xFFFFFF else 0x999999)

        renderTabs(guiGraphics, mouseX, mouseY, px, py)

        // Заголовки колонок
        val headerY = py + headerHeight - 20
        val statsHeaderText = if (mode == Mode.PLAYERS) {
            Component.translatable("krylix.leaderboard.header_kd").string
        } else {
            Component.translatable("krylix.leaderboard.tab_mob_kills").string
        }
        guiGraphics.drawString(font, Component.translatable("krylix.leaderboard.header_rank").string, px + 12, headerY, 0x888888)
        guiGraphics.drawString(font, Component.translatable("krylix.leaderboard.header_player").string, px + avatarColumnX, headerY, 0x888888)
        val statsHeaderWidth = font.width(statsHeaderText)
        guiGraphics.drawString(font, statsHeaderText, px + panelWidth - statsHeaderWidth - 12, headerY, 0x888888)
        guiGraphics.fill(px + 8, py + headerHeight - 4, px + panelWidth - 8, py + headerHeight - 3, 0x40FFFFFF)

        if (fullEntries().isEmpty()) {
            val emptyKey = if (mode == Mode.PLAYERS) "screen.krylix.leaderboard.empty_players" else "screen.krylix.leaderboard.empty_mob_kills"
            val emptyText = Component.translatable(emptyKey).string
            guiGraphics.drawCenteredString(font, emptyText, px + panelWidth / 2, listTop() + listHeight() / 2 - 4, 0xAAAAAA)
        } else {
            renderRows(guiGraphics, mouseX, mouseY)
        }

        val footerText = Component.translatable("krylix.leaderboard.footer", fullEntries().size).string
        guiGraphics.drawCenteredString(font, footerText, px + panelWidth / 2, py + panelHeight - 14, 0x777777)

        super.render(guiGraphics, mouseX, mouseY, partialTick)
    }

    private fun tabRect(index: Int, px: Int, py: Int): IntArray {
        val tabWidth = 90
        val gap = 6
        val totalWidth = tabWidth * 2 + gap
        val startX = px + (panelWidth - totalWidth) / 2
        val x = startX + index * (tabWidth + gap)
        val y = py + 20
        return intArrayOf(x, y, tabWidth, tabHeight)
    }

    private fun renderTabs(guiGraphics: GuiGraphics, mouseX: Int, mouseY: Int, px: Int, py: Int) {
        renderTab(guiGraphics, mouseX, mouseY, tabRect(0, px, py), Component.translatable("krylix.leaderboard.tab_pvp").string, Mode.PLAYERS)
        renderTab(guiGraphics, mouseX, mouseY, tabRect(1, px, py), Component.translatable("krylix.leaderboard.tab_mob_kills").string, Mode.MOB_KILLS)
    }

    private fun renderTab(guiGraphics: GuiGraphics, mouseX: Int, mouseY: Int, rect: IntArray, label: String, tabMode: Mode) {
        val (x, y, w, h) = rect
        val active = mode == tabMode
        val hovered = mouseX in x..(x + w) && mouseY in y..(y + h)
        HudRender.rounded(guiGraphics, x, y, w, h, 3) {
            val bg = when {
                active -> 0xCC3A7BD5.toInt()
                hovered -> 0x30FFFFFF
                else -> 0x20FFFFFF
            }
            guiGraphics.fill(x, y, x + w, y + h, bg)
        }
        val textColor = if (active) 0xFFFFFF else 0xAAAAAA
        val textWidth = font.width(label)
        guiGraphics.drawString(font, label, x + (w - textWidth) / 2, y + (h - 8) / 2, textColor)
    }

    private operator fun IntArray.component1() = this[0]
    private operator fun IntArray.component2() = this[1]
    private operator fun IntArray.component3() = this[2]
    private operator fun IntArray.component4() = this[3]

    private fun renderRows(guiGraphics: GuiGraphics, mouseX: Int, mouseY: Int) {
        val px = panelX()
        val top = listTop()
        val bottom = listBottom()
        prevButtonRect = null
        nextButtonRect = null

        HudRender.rounded(guiGraphics, px + 6, top, panelWidth - 12, bottom - top, 4) {
            val localUuid = minecraft?.player?.uuid?.toString()
            val entries = pageEntries()

            var y = top - scrollOffset
            for (index in entries.indices) {
                if (y + rowHeight >= top && y <= bottom) {
                    val entry = entries[index]
                    val rowHovered = mouseX in (px + 6)..(px + panelWidth - 6) && mouseY in y..(y + rowHeight)
                    val isSelf = entry.uuid == localUuid
                    val bgColor = when {
                        rowHovered -> 0x33FFFFFF
                        isSelf -> 0x2255AAFF.toInt()
                        index % 2 == 0 -> 0x18FFFFFF
                        else -> 0x00000000
                    }
                    if (bgColor != 0) {
                        guiGraphics.fill(px + 6, y, px + panelWidth - 6, y + rowHeight, bgColor)
                    }

                    renderRow(guiGraphics, entry, page * pageSize + index, px, y)
                }
                y += rowHeight
            }

            if (totalPages() > 1 && y + rowHeight >= top && y <= bottom) {
                renderPaginationRow(guiGraphics, mouseX, mouseY, px, y)
            }
        }
    }

    private fun renderPaginationRow(guiGraphics: GuiGraphics, mouseX: Int, mouseY: Int, px: Int, y: Int) {
        val btnSize = 16
        val gap = 8
        val pageText = Component.translatable("krylix.leaderboard.page", page + 1, totalPages()).string
        val pageTextWidth = font.width(pageText)
        val totalWidth = btnSize + gap + pageTextWidth + gap + btnSize
        val startX = px + (panelWidth - totalWidth) / 2
        val btnY = y + (rowHeight - btnSize) / 2

        val hasPrev = page > 0
        val hasNext = page < totalPages() - 1

        val prevRect = intArrayOf(startX, btnY, btnSize, btnSize)
        val nextRect = intArrayOf(startX + btnSize + gap + pageTextWidth + gap, btnY, btnSize, btnSize)
        prevButtonRect = prevRect
        nextButtonRect = nextRect

        renderPageButton(guiGraphics, mouseX, mouseY, prevRect, pointRight = false, hasPrev)
        renderPageButton(guiGraphics, mouseX, mouseY, nextRect, pointRight = true, hasNext)
        guiGraphics.drawString(font, pageText, startX + btnSize + gap, y + (rowHeight - 8) / 2, 0xCCCCCC)
    }

    private fun renderPageButton(guiGraphics: GuiGraphics, mouseX: Int, mouseY: Int, rect: IntArray, pointRight: Boolean, enabled: Boolean) {
        val (x, y, w, h) = rect
        val hovered = enabled && mouseX in x..(x + w) && mouseY in y..(y + h)
        HudRender.rounded(guiGraphics, x, y, w, h, 3) {
            val bg = when {
                !enabled -> 0x15FFFFFF
                hovered -> 0x40FFFFFF
                else -> 0x25FFFFFF
            }
            guiGraphics.fill(x, y, x + w, y + h, bg)
        }
        val color = if (enabled) 0xFFFFFFFF.toInt() else 0xFF555555.toInt()
        renderArrow(guiGraphics, x, y, w, h, pointRight, color)
    }

    /**
     * Рисует стрелку пикселями вместо символов "<"/">" — глиф ">" во встроенном шрифте Minecraft
     * не является зеркальным отражением "<" по метрикам, из-за чего центрирование текстом выглядело
     * визуально смещённым. Треугольник строится геометрически, поэтому обе стрелки гарантированно
     * симметричны друг другу.
     */
    private fun renderArrow(guiGraphics: GuiGraphics, x: Int, y: Int, w: Int, h: Int, pointRight: Boolean, color: Int) {
        val arrowW = 4
        val arrowH = 7
        val originX = x + (w - arrowW) / 2
        val originY = y + (h - arrowH) / 2
        val mid = arrowH / 2
        for (row in 0 until arrowH) {
            val d = kotlin.math.abs(row - mid)
            val length = (arrowW - d).coerceAtLeast(1)
            val rowY = originY + row
            if (pointRight) {
                guiGraphics.fill(originX, rowY, originX + length, rowY + 1, color)
            } else {
                guiGraphics.fill(originX + (arrowW - length), rowY, originX + arrowW, rowY + 1, color)
            }
        }
    }

    private fun renderRow(guiGraphics: GuiGraphics, entry: PlayerStatEntry, rank: Int, px: Int, y: Int) {
        val textY = y + (rowHeight - 8) / 2
        val avatarY = y + (rowHeight - avatarSize) / 2

        val rankText = "#${rank + 1}"
        val rankX = px + rankColumnEnd - font.width(rankText)
        guiGraphics.drawString(font, rankText, rankX, textY, 0xAAAAAA)

        val avatarX = px + avatarColumnX
        val texture = headTextures[entry.uuid]
        HudRender.rounded(guiGraphics, avatarX, avatarY, avatarSize, 2) {
            if (texture != null) {
                RenderSystem.setShaderTexture(0, texture)
                RenderSystem.enableBlend()
                guiGraphics.blit(texture, avatarX, avatarY, avatarSize, avatarSize, 8.0f, 8.0f, 8, 8, 64, 64)
                guiGraphics.blit(texture, avatarX, avatarY, avatarSize, avatarSize, 40.0f, 8.0f, 8, 8, 64, 64)
                RenderSystem.disableBlend()
            } else {
                guiGraphics.fill(avatarX, avatarY, avatarX + avatarSize, avatarY + avatarSize, 0xFF555555.toInt())
            }
        }

        guiGraphics.drawString(font, entry.name, avatarX + avatarSize + 6, textY, 0xFFFFFF)

        val statsText = if (mode == Mode.PLAYERS) {
            val kd = if (entry.deaths == 0) entry.kills.toDouble() else entry.kills.toDouble() / entry.deaths
            "${entry.kills}K / ${entry.deaths}D  (${"%.2f".format(kd)})"
        } else {
            Component.translatable("krylix.leaderboard.mob_kills_count", entry.mobKills).string
        }
        val statsWidth = font.width(statsText)
        guiGraphics.drawString(font, statsText, px + panelWidth - statsWidth - 12, textY, 0xCCCCCC)
    }

    override fun mouseScrolled(mouseX: Double, mouseY: Double, delta: Double): Boolean {
        scrollOffset = (scrollOffset - (delta * rowHeight * 1.5).toInt()).coerceIn(0, maxScroll())
        return true
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        val px = panelX()
        val py = panelY()
        val closeX = px + panelWidth - 18
        val closeY = py + 7
        if (mouseX >= closeX && mouseX <= closeX + 12 && mouseY >= closeY && mouseY <= closeY + 12) {
            onClose()
            return true
        }

        val (t0x, t0y, t0w, t0h) = tabRect(0, px, py)
        if (mouseX >= t0x && mouseX <= t0x + t0w && mouseY >= t0y && mouseY <= t0y + t0h) {
            mode = Mode.PLAYERS
            page = 0
            scrollOffset = 0
            return true
        }
        val (t1x, t1y, t1w, t1h) = tabRect(1, px, py)
        if (mouseX >= t1x && mouseX <= t1x + t1w && mouseY >= t1y && mouseY <= t1y + t1h) {
            mode = Mode.MOB_KILLS
            page = 0
            scrollOffset = 0
            return true
        }

        prevButtonRect?.let { (bx, by, bw, bh) ->
            if (page > 0 && mouseX >= bx && mouseX <= bx + bw && mouseY >= by && mouseY <= by + bh) {
                page--
                scrollOffset = 0
                return true
            }
        }
        nextButtonRect?.let { (bx, by, bw, bh) ->
            if (page < totalPages() - 1 && mouseX >= bx && mouseX <= bx + bw && mouseY >= by && mouseY <= by + bh) {
                page++
                scrollOffset = 0
                return true
            }
        }

        return super.mouseClicked(mouseX, mouseY, button)
    }

    override fun isPauseScreen(): Boolean = false
}
